lib/aiken/bls12_381/scalar.ak

```diff
@@ -0,0 +1,221 @@
+}
+
+/// Constructs a new `Scalar` element from an integer, ensuring it's within the valid range of the field.
+pub fn new(n: Int) -> Option<Scalar> {
+  if n >= 0 && n < bls12_381_scalar_field_prime {
+    Some(Scalar(n))
+  } else {
+    None
+  }
+}
+
+test new_1() {
+  and {
+    new(-1) == None,
+    new(bls12_381_scalar_field_prime) == None,
+    new(834884848) == Some(Scalar(834884848)),
+  }
+}
+
+/// Converts a `Scalar` element back to its integer representation.
+pub fn to_int(self: Scalar) -> Int {
+  self.integer
+}
+
+test to_int_1() {
+  to_int(Scalar(834884848)) == 834884848
+}
+
+/// Adds two `Scalar` elements, ensuring the result stays within the finite field range.
+pub fn add(left: Scalar, right: Scalar) -> Scalar {
+  Scalar(( left.integer + right.integer ) % bls12_381_scalar_field_prime)
+}
+
+test add_1() {
+  and {
+    (add(Scalar(834884848), Scalar(834884848)) == Scalar(1669769696))?,
+    (add(Scalar(bls12_381_scalar_field_prime - 1), Scalar(1)) == Scalar(0))?,
+    (add(
+      Scalar(3),
+      Scalar(
+        52435875175126190479447740508185965837690552500527637822603658699938581184513,
+      ),
+    ) == Scalar(3))?,
+  }
+}
+
+/// Subtracts one `Scalar` element from another, with the result wrapped within the finite field range.
+pub fn sub(left: Scalar, right: Scalar) -> Scalar {
+  Scalar(( left.integer - right.integer ) % bls12_381_scalar_field_prime)
+}
+
+test sub_1() {
+  and {
+    (sub(Scalar(834884848), Scalar(834884848)) == zero())?,
+    (sub(zero(), Scalar(5)) == Scalar(bls12_381_scalar_field_prime - 5))?,
+  }
+}
+
+/// Represents the additive identity (zero) in the `Scalar` field.
+pub fn zero() -> Scalar {
+  Scalar(0)
+}
+
+test zero_1() {
+  (zero() == Scalar(0))?
+}
+
+/// Represents the multiplicative identity (one) in the `Scalar` field.
+pub fn one() -> Scalar {
+  Scalar(1)
+}
+
+test one_1() {
+  one() == Scalar(1)
+}
+
+/// Calculates the additive inverse of a `Scalar` element.
+pub fn neg(self: Scalar) -> Scalar {
+  sub(zero(), self)
+}
+
+test neg_1() {
+  and {
+    neg(Scalar(834884848)) == Scalar(
+      52435875175126190479447740508185965837690552500527637822603658699937746299665,
+    ),
+    neg(zero()) == zero(),
+    neg(one()) == Scalar(bls12_381_scalar_field_prime - 1),
+  }
+}
+
+/// Multiplies two `Scalar` elements, with the result constrained within the finite field.
+pub fn mul(left: Scalar, right: Scalar) -> Scalar {
+  Scalar(left.integer * right.integer % bls12_381_scalar_field_prime)
+}
+
+test mul_1() {
+  and {
+    mul(Scalar(834884848), Scalar(834884848)) == Scalar(697032709419983104),
+    mul(zero(), Scalar(834884848)) == zero(),
+    mul(Scalar(bls12_381_scalar_field_prime - 1), Scalar(2)) == Scalar(
+      52435875175126190479447740508185965837690552500527637822603658699938581184511,
+    ),
+  }
+}
+
+/// Exponentiates an `Scalar` element by a non-negative integer exponent, using repeated squaring.
+pub fn scale(self: Scalar, e: Int) -> Scalar {
```

Comment:
small feedback on scale and power of two. Is it possible to carry out all the multiplication first and then mod at the very end?

Reply:
The expected values for the base value of `scale` are scalars, so integers in `[0,bls12-381-field-prime)`. A typical exponent that one might see in a zkp is of the order of millions. And you are correct that you could carry out the mod later.

But using, for example, a function like the `math.pow` function to calculate the exponent and at the end use mod, would be too expensive (this will not even run on a beefy local machine for a typical example). The reason is that the integer would grow uncontrollably, consume all memory. I think that if we want this efficiently, we need to create a dedicated builtin for modular exponentiation.

The `pow_of_two` function runs efficiently as it only needs `k` multiplications in the field (it's less than 1% of the cpu usage for a typical example).

But maybe it is better to not expose (or delete) the function `scale`/`div`/`recip` since they are not that usable onchain (~10% of the cpu budget). An argument to hide them and not delete them is to show with tests that the implementation is correct.